src/HTTP.jl

```diff
@@ -323,14 +327,90 @@ end
 """
 function request(method, url, h=Header[], b=nobody;
                  headers=h, body=b, query=nothing, kw...)::Response
-    return request(HTTP.stack(;kw...), string(method), request_uri(url, query), mkheaders(headers), body; kw...)
+    return request(HTTP.stack(), method, url, headers, body, query; kw...)
 end
-function request(stack::Type{<:Layer}, method, url, h=Header[], b=nobody;
-                 headers=h, body=b, query=nothing, kw...)::Response
-    return request(stack, string(method), request_uri(url, query), mkheaders(headers), body; kw...)
+
+const STREAM_LAYERS = [timeoutlayer, exceptionlayer, debuglayer]
+const REQUEST_LAYERS = [messagelayer, redirectlayer, defaultheaderslayer, basicauthlayer, contenttypedetectionlayer, cookielayer, retrylayer, canonicalizelayer]
+
+pushlayer!(layer; request::Bool=true) = push!(request ? REQUEST_LAYERS : STREAM_LAYERS, layer)
+pushfirstlayer!(layer; request::Bool=true) = pushfirst!(request ? REQUEST_LAYERS : STREAM_LAYERS, layer)
+poplayer!(; request::Bool=true) = pop!(request ? REQUEST_LAYERS : STREAM_LAYERS)
+popfirstlayer!(; request::Bool=true) = popfirst!(request ? REQUEST_LAYERS : STREAM_LAYERS)
+
+function stack(
+    # custom layers
+    requestlayers=(),
+    streamlayers=())
+
+    # stream layers
+    layers = foldr((x, y) -> x(y), streamlayers, init=streamlayer)
+    layers2 = foldr((x, y) -> x(y), STREAM_LAYERS, init=layers)
+    # request layers
+    # messagelayer must be the 1st/outermost layer to convert initial args to Request
+    layers3 = foldr((x, y) -> x(y), requestlayers; init=connectionlayer(layers2))
+    return foldr((x, y) -> x(y), REQUEST_LAYERS; init=layers3)
+end
+
+function request(stack::Base.Callable, method, url, h=Header[], b=nobody, q=nothing;
+                 headers=h, body=b, query=q, kw...)::Response
+    return stack(string(method), request_uri(url, query), mkheaders(headers), body; kw...)
 end
 
-request(::Type{Union{}}, resp::Response) = resp
+macro client(requestlayers, streamlayers=[])
```

Comment:
New lexically-scoped client macro generator that provides the most common/convenient user-facing functions that can be defined in your own module.